### ISSUE TITLE
docker ps -l doesn't list the existing container

### ISSUE THREAD

**Issue by lalyos** [docker/cli#2446]
**Description**

I have removed a lot of stopped container, except one. Then I was trying to list the only existing (stopped) container with `docker ps -l` or `docker ps -n1` but it yields nothing.
```
$ docker ps -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
$ docker ps -n1
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

On the contrary `docker ps -a` does lists the container:
```
$ docker ps -qa
8f0b52e5b059
```

Strangely if I increase the argument for `-n/--last` the container is listed:
```
$ docker ps -qn 1
$ docker ps -qn 2
$ docker ps -qn 10
$ docker ps -qn 11
$ docker ps -qn 12
$ docker ps -qn 13
8f0b52e5b059
```

**Steps to reproduce the issue:**

**Describe the results you received:**

```
> docker ps -ql
# nothing
```

**Describe the results you expected:**

```
> docker ps -ql
8f0b52e5b059
```

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:21:11 2020
 OS/Arch:           darwin/amd64
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:29:16 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false
 Plugins:
  app: Docker Application (Docker Inc., v0.8.0)
  buildx: Build with BuildKit (Docker Inc., v0.3.0-5-g5b97415-tp-docker)
  clip: Docker Client Plugins Manager (Łukasz Lach, v19.06.0)
  dive: Explore contents of the image layers (Łukasz Lach, v19.06.0)
  microscan: Scan the security of an image with Aqua Microscanner (Łukasz Lach, v19.06.0)
  runlike: Get the command line necessary to run a copy of any container (Łukasz Lach, v19.06.0)
  wasm: Run wasm containers (Docker Inc.)
  mutagen: Synchronize files with Docker Desktop (Docker Inc., testing)

Server:
 Containers: 13
  Running: 11
  Paused: 0
  Stopped: 2
 Images: 11
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 1.947GiB
 Name: docker-desktop
 ID: QNMK:EWUP:NUXS:ICP5:CKGQ:AVYG:AEDB:XLSX:LK3Y:G3C7:GNBV:5GUO
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 102
  Goroutines: 96
  System Time: 2020-04-16T08:26:08.3063448Z
  EventsListeners: 3
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  docker.for.mac.localhost:5000
  localhost:5000
  registry.ing.k8z.eu
  registry.k8z.eu
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
Strange; what's the state of the last container, and of previous containers?

Trying to reproduce (I was wondering if it would be somehow related to "state" (last container running or not)), but I'm not able to reproduce :thinking_face:


First with two containers running, the last container being stopped:

```bash
docker run -dit --name running-container busybox
docker run --name stopped-container busybox

docker ps -a

CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
91a3ca693f6a        busybox             "sh"                11 seconds ago      Exited (0) 6 seconds ago                       stopped-container
01ce407487b6        busybox             "sh"                15 seconds ago      Up 14 seconds                                  running-container

docker ps -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                      PORTS               NAMES
91a3ca693f6a        busybox             "sh"                28 seconds ago      Exited (0) 23 seconds ago                       stopped-container

docker ps -n1
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                      PORTS               NAMES
91a3ca693f6a        busybox             "sh"                50 seconds ago      Exited (0) 45 seconds ago                       stopped-container

docker ps -n2
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS                          PORTS               NAMES
91a3ca693f6a        busybox             "sh"                About a minute ago   Exited (0) About a minute ago                       stopped-container
01ce407487b6        busybox             "sh"                About a minute ago   Up About a minute                                   running-container
```

Checking if the `-q` / `--quiet` option would play a role:

```bash
docker ps -qn1
91a3ca693f6a

docker ps -qn2
91a3ca693f6a
01ce407487b6
```

And trying if it reproduces if the last container is running (starting a new container that keeps running);

```bash
docker run -dit --name running-container-two busybox

docker ps -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
7092212703d6        busybox             "sh"                9 seconds ago       Up 9 seconds                            running-container-two

docker ps -n1
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
7092212703d6        busybox             "sh"                31 seconds ago      Up 31 seconds                           running-container-two
```

**Comment by thaJeztah** [comment]
oh; let me add the version on which I tested; this is on Docker for Mac, latest version;

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:21:11 2020
 OS/Arch:           darwin/amd64
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:29:16 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by lalyos** [comment]
Indeed very strange. The last existing container is in the exited state:
```
docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                   PORTS               NAMES
8f0b52e5b059 
```

But previously there was a lot of other exited  containers (probably 12), which I removed.

It seems to me that the daemon has an outdated snaphot of containers:
https://github.com/moby/moby/blob/7d0e9399a357e5a897db4ede8d39d91d678b9b64/daemon/list.go#L184

This last container was probably really the 13th container, before I removed all but this.

## Further experiment

If I start a new container, than I have 2 container the:
- the first (the new one)
- and the other one, now  on position 14

```
$ docker ps -qn1
# nothing
docker ps -qn13
8f0b52e5b059
# container 8f0b is now on position 13

$ docker run -d --name delme  nginx:alpine
d6a8f10157f0cfcf303c03d4045cf3000702a077181aa00c7e3b044f76f03446
$ docker ps -qn1
d6a8f10157f0
$ docker ps -qn2
d6a8f10157f0
$ docker ps -qn10
d6a8f10157f0
$ docker ps -qn13
d6a8f10157f0
$ docker ps -qn14
d6a8f10157f0
8f0b52e5b059
```

I have now 2 container the 1th and the 14th ???

**Comment by lalyos** [comment]
## Experiment 2

I was checking if a combination of containers, before the removal could make a difference:
- exited
- exited
- runing
- exited

But still couldn't reproduce:
```
$ docker run --name first alpine echo 1
1
$ docker run -d  --name second nginx:alpine
7c417dc3eb46aa1d3553433462b271789cdf6b5449c883b7ded5b9ce954556ea
$ docker run alpine echo 3
3
$ docker run alpine echo 4
4
$
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                      PORTS               NAMES
2ad8b1ab5e86        alpine              "echo 4"                 11 seconds ago      Exited (0) 9 seconds ago                        wizardly_roentgen
71880e2504d9        alpine              "echo 3"                 12 seconds ago      Exited (0) 11 seconds ago                       xenodochial_ishizaka
7c417dc3eb46        nginx:alpine        "nginx -g 'daemon of…"   14 seconds ago      Up 12 seconds               80/tcp              second
20c945f97e4f        alpine              "echo 1"                 15 seconds ago      Exited (0) 14 seconds ago                       first
$ docker rm -f $(docker ps -qa --filter=since=first)
2ad8b1ab5e86
71880e2504d9
7c417dc3eb46
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                      PORTS               NAMES
20c945f97e4f        alpine              "echo 1"            27 seconds ago      Exited (0) 26 seconds ago                       first
$ docker ps -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                      PORTS               NAMES
20c945f97e4f        alpine              "echo 1"            28 seconds ago      Exited (0) 27 seconds ago                       first
```

**Comment by lalyos** [comment]
## Experiment 3

I tried to create multiple blocks of running/exited containers after the first one, and then delete all but the first:
```
$ for n in {1..5}; do docker run alpine echo $n; done
$ for n in {1..5}; do docker run -d  nginx:alpine; done
$ for n in {6..10}; do docker run alpine echo $n; done
$ for n in {1..5}; do docker run -d  nginx:alpine; done
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED              STATUS                          PORTS               NAMES
c44eeacccbb9        nginx:alpine        "nginx -g 'daemon of…"   5 seconds ago        Up 3 seconds                    80/tcp              elated_shirley
b57c3943d81c        nginx:alpine        "nginx -g 'daemon of…"   6 seconds ago        Up 4 seconds                    80/tcp              elastic_cartwright
58255aecfc81        nginx:alpine        "nginx -g 'daemon of…"   7 seconds ago        Up 5 seconds                    80/tcp              recursing_fermi
5981ac82f824        nginx:alpine        "nginx -g 'daemon of…"   9 seconds ago        Up 7 seconds                    80/tcp              agitated_ishizaka
40fed6d15bbb        nginx:alpine        "nginx -g 'daemon of…"   10 seconds ago       Up 8 seconds                    80/tcp              objective_herschel
66429c621dd0        alpine              "echo 10"                20 seconds ago       Exited (0) 18 seconds ago                           epic_thompson
8f7a57f4c22c        alpine              "echo 9"                 22 seconds ago       Exited (0) 20 seconds ago                           hungry_bell
47f2179b5c7b        alpine              "echo 8"                 24 seconds ago       Exited (0) 21 seconds ago                           frosty_matsumoto
7b289b465d67        alpine              "echo 7"                 25 seconds ago       Exited (0) 23 seconds ago                           wonderful_ardinghelli
2893e94b8bb9        alpine              "echo 6"                 27 seconds ago       Exited (0) 25 seconds ago                           relaxed_volhard
f956c896c2d9        nginx:alpine        "nginx -g 'daemon of…"   46 seconds ago       Up 44 seconds                   80/tcp              zen_brattain
04ca55993e73        nginx:alpine        "nginx -g 'daemon of…"   47 seconds ago       Up 45 seconds                   80/tcp              nostalgic_bhaskara
a8a8e27906fd        nginx:alpine        "nginx -g 'daemon of…"   48 seconds ago       Up 46 seconds                   80/tcp              xenodochial_elion
f81ed6072a01        nginx:alpine        "nginx -g 'daemon of…"   49 seconds ago       Up 47 seconds                   80/tcp              naughty_gagarin
d8ac8a87e71e        nginx:alpine        "nginx -g 'daemon of…"   50 seconds ago       Up 48 seconds                   80/tcp              mystifying_chatterjee
dc7982320a9a        alpine              "echo 5"                 About a minute ago   Exited (0) About a minute ago                       elegant_shockley
f4335e2c23e0        alpine              "echo 4"                 About a minute ago   Exited (0) About a minute ago                       brave_cartwright
30d12dd08d7a        alpine              "echo 3"                 About a minute ago   Exited (0) About a minute ago                       elated_brown
23873ec10bcf        alpine              "echo 2"                 About a minute ago   Exited (0) About a minute ago                       gallant_wescoff
bdd06d31c74a        alpine              "echo 1"                 About a minute ago   Exited (0) About a minute ago                       angry_hamilton
20c945f97e4f        alpine              "echo 1"                 6 minutes ago        Exited (0) 6 minutes ago                            first
$ docker rm -f $(docker ps -qa --filter=since=first)
$ docker ps -l
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
20c945f97e4f        alpine              "echo 1"            9 minutes ago       Exited (0) 9 minutes ago                       first
```

still no cigar ...

**Comment by thaJeztah** [comment]
Thanks for trying to provide more details. Thats a pity that it's not consistently reproducible. Definitely possible there's a bug/race condition/syncing issue somewhere (at this point I don't have a direct idea _what_ the cause could be).

Were all hosts running the same version of the engine? And are all hosts Docker for Mac, or also different kind of setups? Docker Desktop uses an internal proxy for connections with the daemon, and we've had cases where a bug was in that part, so trying to exclude that as possible cause.

**Comment by lalyos** [comment]
## Reproduced

Oh i figured: yesterday I was playing with k8s (with k3s specifically), so what really happens is that there are really 12 other containers, but their are hidden from outside (from mac host).

If I run the same docker `ps -a` command in linuxkit I can see those containers:

```
$ docker ps -qa
8f0b52e5b059

$ docker run -it --rm --privileged --pid=host alpine nsenter -t 1 -m -u -n -i -- sh -c "ctr -n services.linuxkit task exec -t --exec-id 5555 docker docker ps -qa "
365677ad9d17
d8adf5a50b0c
d60dfcc24209
1d8addfe677d
22584fb6c447
8979eefc9906
89d09ad39c8b
f996c2231e3a
0bbd86aa5843
e34cc43af8ef
2aee1c77f368
2fb6ad9d8c98
e2c87a3cf893
eb19b2fa1a9b
8f0b52e5b059
```

So I can create a hidden container by:
```
$ docker run -d \
  --label io.kubernetes.docker.type=podsandbox\
  --name=hidden \
  nginx:alpine                                                                                                                                                    
8114cad0ab573d7f1536e2369a9b33d1c49b8ec1b95dd4992fcfc8c9226a2ba6                                                                                                                                                                            
$ docker ps -a                                                                                                                                                                                                                              
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                   PORTS               NAMES                                                                                                     
8f0b52e5b059        debian              "nsenter -t 1224 -m …"   3 days ago          Exited (1) 7 hours ago                       intelligent_elion                                                                                         
# 8114 is hidden
# but I can inspect it from Mac host:
$ docker inspect 8114                                                                                                                                                                                                                       
[                                                                                                                                                                                                                                           
    {                                                                                                                                                                                                                                       
        "Id": "8114cad0ab573d7f1536e2369a9b33d1c49b8ec1b95dd4992fcfc8c9226a2ba6",                                                                                                                                                           
        "Created": "2020-04-16T12:55:19.7593328Z",                                                                                                                                                                                          
        "Path": "nginx",
...
]
```

**Comment by lalyos** [comment]
![Screenshot 2020-04-16 at 14 59 39](https://user-images.githubusercontent.com/464122/79459000-eb9d7700-7ff2-11ea-8612-a3fae59b8d85.png)

When k8s is enabled via docker/preferences/kubernetes, I have a choise to "Show Kubernetes internal containers when using Docker commands.", but when k8s is disabled, this checkbox is disabled.

Is this something I could configure in ~/.docker/config.json or somwhere else?

**Comment by lalyos** [comment]
![Screenshot 2020-04-16 at 15 03 10](https://user-images.githubusercontent.com/464122/79459418-89914180-7ff3-11ea-894b-65ccb92308bb.png)

I start from a disabled k8s state. I can enable to see _all_ containers, even without starting k8s:
- Go to docker/prefernces/kubernetes
- click on "Enable Kubernetes"
- click on "Show system containers (advanced)"
- click again on "Enable Kubernetes" (to disable again)
- "apply & restart"

It will not restart the daemon, but will enable to see all (k8s labeled) containers.

**Comment by lalyos** [comment]
So it's not a bug, thanks @thaJeztah for the quick response and valuable tips. Does it needs some documentation enhancement (I can create a PR) or its just "common sense" ;) and I can close this?

**Comment by thaJeztah** [comment]
oh! good call, yes, I think this is something Docker Desktop handles and automatically applies a filter based on those labels 🤔 


@djs55 @nebuk89 do we have an issue tracking this? I can see this being problematic in some cases (I also use the `-n1` / `-l` option at times to, e.g. `docker exec` into the last container I created).

**Comment by thaJeztah** [comment]
Actually; just thinking now; if Docker Desktop uses the equivalent of `docker ps --filter label=foo!=kubernetes`, then perhaps it can be solved by _first_ applying the label filter, and after that picking the last container.

Question is if that's "correct" (at least it would be a change in behaviour) 🤔

**Comment by lalyos** [comment]
Actually this "Show system containers" is changing the client or the daemon side behaviour?
As I cloudn't find any `podsandbox` mention in moby/moby or docker/docker-cli. 

I also found a workaround to see the unfiltered container list, regardless of the docker/preferences/kubernetes setings:

```
docker  \
  --host unix:///$HOME/Library/Containers/com.docker.docker/Data/docker.raw.sock \
  ps
```
I guess this what you referred as:
> Docker Desktop uses an internal proxy for connections with the daemon
The proxy is /Users/lalyos/Library/Containers/com.docker.docker/Data/docker.sock
while the direct socket is docker.raw.sock ?


Btw is docker-desktop.iso  built from: https://github.com/linuxkit/linuxkit/blob/master/examples/docker-for-mac.yml
I couldn't find out how docker-for-mac is built. I remember something that docker-desktop source was open on github, but I can't find it anymore.  Isn't it open source?

**Comment by nebuk89** [comment]
woops missed this! @lalyos that is an interesting but 🤔 I am unsure how many people have encountered it before but I get how it can be confusing! I think docs as you highlight is the best approach, I can tag this to @usha-mandya or if you want to open a PR against our docs repo and tag me we can take a look :) 

and to answer your other question, that is the old yml that mac used, all of this is now closed source rather than open (and the main desktop codebase was never open). 

Thanks!

**Comment by abdelghanimeliani** [comment]
i had the same provblem this morning

**Comment by KerickHowlett** [comment]
## TL;DR

Enter the following command:

```bash
docker context use default
```

## Full Explanation

### The Low-Level Problem

Having just experiencedd the same problem myself, I went off a hunch and soon discovered file located at the following file path (that will be different for you but also similar enough):

```text
$HOME/.docker/contexts/meta/fe9c6bd7a66301f49ca9b6a70b217107cd1284598bfc254700c989b916da791e/meta.json
```

This JSON file contains the following:

```json
{
	"Name": "desktop-linux",
	"Metadata": {},
	"Endpoints": {
		"docker": {
			"Host": "unix:///Users/USERNAME/.docker/run/docker.sock",
			"SkipTLSVerify": false
		}
	}
}
```

The `Endpoints.docker.Host` path is what dictates the path to the Docker daemon, so it's possible you could also apply @lalyos' workaround by pasting the UNIX socket path in place of that property's value.

### Another Possible Workaround

Although, as you might imagine with the hashed directory name within the file path, this isn't a permanent file, and chances are high that your changes will be overridden the next time you reboot your Mac device, restart Docker, and/or update Docker Desktop.

You could always create a shell script to be placed within your `.bashrc`, `.zshrc`, or `$__fish_config_dir`.

### The _Actual_ Solution

While it's still just a theory at this point, I think I may have discovered the culprit to this particular blight:

If you enter enter `docker context ls` into your terminal, I'm willing to bet you'll see at least TWO contexts listed - `default` and `desktop-linux` -- with the latter selected.

Assuming I'm correct, if you just enter `docker context use default`, it should update the `Endpoints.docker.Host` property within your local `meta.json` file with the correct path to your Docker Desktop daemon.

### The High-Level Cause

This type of thing just doesn't happen on its own, and if you're like me, you don't mess with Docker's context feature all that much; however, this kinda thing doesn't just happen on its own, though, right?

Well, I believe I have another theory as to what may be the root cause:

If you're like me, a developer who uses the Docker extension for Visual Studio Code, and have your Visual Studio Code profile configured to sync-up across all your machines -- at least one of which is running some flavor Linux (e.g., Debian or Ubuntu) and at least one other is running MacOS.

### The Long-Term Solution

Unfortunately, at the time of this post, there isn't a native way to configure a default Docker context in your Visual Studio Code's `settings.json` file in any way shape or form -- let alone a way to configure it based on Operating System.

Although, that may not be needed. It may be enough for the Docker development team (or some unsung open-source hero who's well versed in Golang) to apply some sort of verification check based on the active Operating System to make sure the correct daemon is used (assuming that isn't already implemented).

Again, this is just a theory on my part.
